tsl/src/chunk_api.c

```diff
@@ -371,3 +372,126 @@ chunk_create(PG_FUNCTION_ARGS)
 
 	PG_RETURN_DATUM(HeapTupleGetDatum(tuple));
 }
+
+
+/*
+ * Detach a chunk from a hypertable.
+ */
+Datum
+chunk_detach(PG_FUNCTION_ARGS)
+{
+	Oid chunk_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Cache *hcache;
+	Hypertable *ht;
+	Chunk *chunk;
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, true);
+	if (!chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_TABLE),
+				 errmsg("chunk with OID %u does not exist", chunk_relid)));
+
+	ht = ts_hypertable_cache_get_cache_and_entry(chunk->hypertable_relid, CACHE_FLAG_NONE, &hcache);
+	Assert(NULL != ht);
+
+	if (!object_ownercheck(RelationRelationId, chunk_relid, GetUserId()))
+		aclcheck_error(ACLCHECK_NOT_OWNER,
+					   get_relkind_objtype(chunk->relkind),
+					   get_rel_name(chunk_relid));
+
+	if (ts_chunk_is_compressed(chunk))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot detach compressed chunk \"%s\"", get_rel_name(chunk_relid)),
+				 errhint("decompress the chunk first")));
+
+	if (chunk->fd.osm_chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot detach OSM chunk \"%s\"", get_rel_name(chunk_relid))));
+
+	LockRelationOid(chunk->hypertable_relid, ShareUpdateExclusiveLock);
+	LockRelationOid(chunk->table_id, AccessExclusiveLock);
+
+	AlterTableCmd cmd = {
+		.type = T_AlterTableCmd,
+		.subtype = AT_DropInherit,
+		.def = (Node *) makeRangeVar(NameStr(ht->fd.schema_name), NameStr(ht->fd.table_name), 0),
+	};
+	AlterTableStmt stmt = {
+		.type = T_AlterTableStmt,
+		.cmds = list_make1(&cmd),
+		.relation = makeRangeVar(NameStr(ht->fd.schema_name), NameStr(ht->fd.table_name), 0),
+	};
+
+	ts_alter_table_with_event_trigger(chunk->table_id, (Node *) &stmt, list_make1(&cmd), false);
+
+	ts_chunk_detach_by_relid(chunk_relid);
+
+	ts_cache_release(&hcache);
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * Attach an existing relation to a hypertable as a chunk.
+ */
+Datum
+chunk_attach(PG_FUNCTION_ARGS)
+{
+	Oid ht_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	Jsonb *slices = PG_ARGISNULL(2) ? NULL : PG_GETARG_JSONB_P(2);
+	Cache *hcache;
+	Hypertable *ht;
+	Hypercube *hc;
+	Chunk *chunk;
+	Relation chunk_rel;
+	bool created;
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	check_privileges_for_creating_chunk(ht_relid);
+	ht = ts_hypertable_cache_get_cache_and_entry(ht_relid, CACHE_FLAG_NONE, &hcache);
+	Assert(NULL != ht);
+
+	if (NULL == slices)
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("invalid slices")));
+	hc = get_hypercube_from_slices(slices, ht);
+	Assert(NULL != hc);
```

Comment:
Maybe you want to do this after you got the lock on the hypertable. Otherwise there is a (granted, very small) window where privileges and/or slices might change before you have a chance to get the lock on the hypertable.